kevm-pyk/src/kevm_pyk/solc_to_k.py

```diff
@@ -294,72 +294,73 @@ def solc_compile(contract_file: Path, profile: bool = False) -> Dict[str, Any]:
+    module = KFlatModule(module_name, contract.sentences, [KImport(i) for i in ['EDSL'] + list(imports)])
 
-    claims_module: Optional[KFlatModule] = None
-    contract_function_application_label = contract.klabel_method
-    function_test_calldatas = []
-    for tm in contract.methods:
-        if tm.name.startswith('test'):
-            klabel = tm.production.klabel
-            assert klabel is not None
-            args = [
-                abstract_term_safely(KVariable('_###SOLIDITY_ARG_VAR###_'), base_name=f'V{name}')
-                for name in tm.arg_names
-            ]
-            calldata: KInner = KApply(
-                contract_function_application_label, [KApply(contract.klabel), KApply(klabel, args)]
-            )
-            callvalue: KInner = (
-                intToken(0)
-                if not tm.payable
-                else abstract_term_safely(KVariable('_###CALLVALUE###_'), base_name='CALLVALUE')
-            )
-            function_test_calldatas.append((tm.name, calldata, callvalue))
-    if function_test_calldatas:
-        claims = _gen_claims_for_contract(empty_config, contract.name, calldata_cells=function_test_calldatas)
-        import_module = main_module if main_module else module_name
-        claims_module = KFlatModule(module_name + '-SPEC', claims, [KImport(import_module)])
+    test_methods = [method for method in contract.methods if method.name.startswith('test')]
+    if not test_methods:
+        return module, None
 
+    claims = [_test_execution_claim(empty_config, contract, method) for method in test_methods]
+    import_module = main_module if main_module else module_name
+    claims_module = KFlatModule(module_name + '-SPEC', claims, [KImport(import_module)])
     return module, claims_module
 
 
-def _gen_claims_for_contract(
-    empty_config: KInner, contract_name: str, calldata_cells: List[Tuple[str, KInner, KInner]] = None
-) -> List[KClaim]:
+def _test_execution_claim(empty_config: KInner, contract: Contract, method: Contract.Method) -> KClaim:
+    claim_name = method.name.replace('_', '-')
+    calldata = _calldata_cell(contract, method)
+    callvalue = _callvalue_cell(method)
+    init_term = _init_term(empty_config, contract.name)
+    init_cterm = _init_cterm(_init_term_with_calldata(init_term, calldata, callvalue))
+    failing = method.name.startswith('testFail')
+    final_cterm = _final_cterm(empty_config, contract.name, failing=failing)
+    claim, _ = build_claim(claim_name, init_cterm, final_cterm)
+    return claim
+
+
+def _calldata_cell(contract: Contract, method: Contract.Method) -> KInner:
+    contract_function_application_label = contract.klabel_method
+    klabel = method.production.klabel
+    assert klabel is not None
+    args = [
+        abstract_term_safely(KVariable('_###SOLIDITY_ARG_VAR###_'), base_name=f'V{name}') for name in method.arg_names
+    ]
+    return KApply(contract_function_application_label, [KApply(contract.klabel), KApply(klabel, args)])
```

Comment:
```suggestion
def _calldata_cell(contract: Contract, method: Contract.Method) -> KInner:
    contract_function_application_label = contract.klabel_method
    return KApply(contract_function_application_label, [KApply(contract.klabel), method.application])
``` 

Then this ^^^ can likely just be inlined where it is used.

Then add to class `Contract.Method`:

```
@property
def application(self: Contract.Method) -> KApply:
    klabel = self.production.klabel
    assert klabel is not None
    args = [
        abstract_term_safely(KVariable('_###SOLIDITY_ARG_VAR###_'), base_name=f'V{name}') for name in self.arg_names
    ]
    return KApply(klabel, args)
```

Reply:
Done.